kevlar/include/kevlar_bits/model/binomial/berry_inla.hpp

```diff
@@ -0,0 +1,368 @@
+#pragma once
+#include <Eigen/Dense>
+#include <algorithm>
+#include <kevlar_bits/model/base.hpp>
+#include <kevlar_bits/model/binomial/common/fixed_n_default.hpp>
+#include <kevlar_bits/model/fixed_single_arm_size.hpp>
+#include <kevlar_bits/util/algorithm.hpp>
+#include <kevlar_bits/util/legendre.hpp>
+#include <kevlar_bits/util/macros.hpp>
+#include <kevlar_bits/util/math.hpp>
+#include <kevlar_bits/util/types.hpp>
+
+namespace kevlar {
+namespace model {
+namespace binomial {
+
+template <class ValueType, int ARMS>
+struct BerryINLA : FixedSingleArmSize, ModelBase<ValueType> {
+    using arm_base_t = FixedSingleArmSize;
+    using base_t = ModelBase<ValueType>;
+    using typename base_t::value_t;
+
+   private:
+    using vec_t = colvec_type<value_t>;
+    using mat_t = mat_type<value_t>;
+
+    static constexpr double mu_0 = -1.34;
+
+    const size_t n_arm_samples;
+    const vec_t efficacy_thresholds;
+
+    // variables dependent solely on the quadrature points (independent of the
+    // data).
+    const vec_t quad_wts;
+    const mat_t cov;
+    const mat_t neg_precQ;
+    const vec_t logprecQdet;
+    const vec_t logprior;
+    const value_t opt_tol2;
+    const value_t logit_p1;
+
+   public:
+    template <class _GenType, class _ValueType, class _UIntType,
+              class _GridRangeType>
+    struct SimGlobalState;
+
+    template <class _GenType, class _ValueType, class _UIntType,
+              class _GridRangeType>
+    using sim_global_state_t =
+        SimGlobalState<_GenType, _ValueType, _UIntType, _GridRangeType>;
+
+    template <class _GridRangeType>
+    using kevlar_bound_state_t = KevlarBoundStateFixedNDefault<_GridRangeType>;
+
+    BerryINLA(size_t n_arm_samples, const Eigen::Ref<const vec_t>& cv,
+              const Eigen::Ref<const vec_t>& efficacy_thresholds,
+              const Eigen::Ref<const vec_t>& quad_wts,
+              const Eigen::Ref<const mat_t>& cov,
+              const Eigen::Ref<const mat_t>& neg_precQ,
+              const Eigen::Ref<const vec_t>& logprecQdet,
+              const Eigen::Ref<const vec_t>& logprior, value_t opt_tol,
+              value_t logit_p1)
+        : arm_base_t(ARMS, n_arm_samples),
+          base_t(),
+          n_arm_samples(n_arm_samples),
+          efficacy_thresholds(efficacy_thresholds),
+          quad_wts(quad_wts),
+          cov(cov),
+          neg_precQ(neg_precQ),
+          logprecQdet(logprecQdet),
+          logprior(logprior),
+          opt_tol2(opt_tol * opt_tol),
+          logit_p1(logit_p1) {
+        assert(efficacy_thresholds.size() == ARMS);
+        set_critical_values(cv);
+    }
+
+    using base_t::critical_values;
+    void set_critical_values(const Eigen::Ref<const colvec_type<value_t>>& cv) {
```

Comment:
Can this function be inlined because it's so short?

Reply:
Yeah, seems nice!